src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs

```diff
@@ -0,0 +1,194 @@
+        /// <summary>
+        /// TryCreate method to create a new object.
+        /// </summary>        
+        /// <param name="keyValues">The key-value pair of the object to be created. Optional</param>
+        /// <param name="createdObject">The created object (CLR or Typeless)</param>
+        /// <param name="errorMessage">Any error message in case of an exception</param>
+        /// <returns>The status of the TryCreate method <see cref="ODataAPIResponseStatus"/> </returns>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1021:AvoidOutParameters", MessageId = "2#")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1021:AvoidOutParameters", MessageId = "1#")]
+        public abstract ODataAPIResponseStatus TryCreate(IDictionary<string, object> keyValues, out TStructuralType createdObject, out string errorMessage);
+
+        /// <summary>
+        /// TryGet method which tries to get the original object based on key-values.
+        /// </summary>
+        /// <param name="keyValues">Key-value pair for the entity keys.</param>
+        /// <param name="originalObject">Object to return.</param>
+        /// <param name="errorMessage">Any error message in case of an exception.</param>
+        /// <returns>The status of the TryGet method <see cref="ODataAPIResponseStatus"/>.</returns>
+        public abstract ODataAPIResponseStatus TryGet(IDictionary<string, object> keyValues, out TStructuralType originalObject, out string errorMessage);
+
+        /// <summary>
+        /// TryDelete Method which will delete the object based on key-value pairs.
+        /// </summary>
+        /// <param name="keyValues">Key-value pair for the entity keys.</param>
+        /// <param name="errorMessage">Any error message in case of an exception.</param>
+        /// <returns>The status of the TryGet method <see cref="ODataAPIResponseStatus"/>.</returns>
+        public abstract ODataAPIResponseStatus TryDelete(IDictionary<string, object> keyValues, out string errorMessage);
+
+        /// <summary>
+        /// Get the ODataAPIHandler for the nested type.
+        /// </summary>
+        /// <param name="parent">Parent instance.</param>
+        /// <param name="navigationPropertyName">The name of the navigation property for the handler.</param>
+        /// <returns>The type of Nested ODataAPIHandler.</returns>
+        public abstract IODataAPIHandler GetNestedHandler(TStructuralType parent, string navigationPropertyName);
+
+        /// <summary>
+        /// Add related object.
+        /// </summary>
+        /// <param name="resource">The object to be added.</param>
+        /// <param name="errorMessage">Any error message in case of an exception.</param>
+        /// <returns>The status of the AddRelatedObject method <see cref="ODataAPIResponseStatus"/>.</returns>
+        public abstract ODataAPIResponseStatus TryAddRelatedObject(TStructuralType resource, out string errorMessage);
```

Comment:
I think the parent object needs to be passed to this method..

Reply:
Resolved offline